src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,405 @@
+
+	buf_size = cd->chunk_frames * sizeof(cd->input[0]) * cd->num_channels;
+	cd->input = rballoc(0, SOF_MEM_CAPS_RAM, buf_size);
+	if (!cd->input) {
+		ret = -ENOMEM;
+		goto fail;
+	}
+	cd->output = rballoc(0, SOF_MEM_CAPS_RAM, buf_size);
+	if (!cd->output) {
+		ret = -ENOMEM;
+		goto fail;
+	}
+
+	cd->input_samples = 0;
+	cd->next_avail_output_samples = cd->chunk_frames * cd->num_channels;
+
+	cd->state = GoogleCtcAudioProcessingCreateWithConfig(cd->num_frames,
+			CONFIG_COMP_GOOGLE_CTC_AUDIO_PROCESSING_PARTITION_SIZE,
+			CONFIG_COMP_GOOGLE_CTC_AUDIO_PROCESSING_IMPULSE_SIZE,
+			cd->chunk_frames,
+			CONFIG_COMP_GOOGLE_CTC_AUDIO_PROCESSING_SAMPLE_RATE,
+			/*is_symmetric=*/0, /*config=*/NULL, /*config_size=*/0);
+
+	cd->tuning_handler = comp_data_blob_handler_new(dev);
+	if (!cd->tuning_handler) {
+		ret = -ENOMEM;
+		goto fail;
+	}
+
+	comp_dbg(dev, "ctc_init(): Ready");
+
+	return 0;
+fail:
+	comp_err(dev, "ctc_init(): Failed");
+	ctc_free(mod);
+	return ret;
+}
+
+static int google_ctc_audio_processing_reconfigure(struct processing_module *mod)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	uint8_t *config;
+	size_t size;
+	int ret;
+
+	comp_dbg(dev, "google_ctc_audio_processing_reconfigure()");
+
+	config = comp_get_data_blob(cd->tuning_handler, &size, NULL);
+	if (size == 0) {
+		/* No data to be handled */
+		return 0;
+	}
+
+	if (!config) {
+		comp_err(dev, "google_ctc_audio_processing_reconfigure(): Tuning config not set");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
+		  config, size);
+
+	cd->reconfigure = false;
+
+	uint8_t *processing_config;
+	size_t processing_config_size;
+	bool processing_config_present;
+
+	GoogleCtcAudioProcessingParseSofConfigMessage(config, size,
+						      &processing_config,
+						      &processing_config_size,
+						      &processing_config_present);
+
+	if (processing_config_present) {
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying config of size %zu bytes",
+			  processing_config_size);
+
+		ret = GoogleCtcAudioProcessingReconfigure(cd->state,
+							  processing_config,
+							  processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int ctc_prepare(struct processing_module *mod,
+		       struct sof_source **sources, int num_of_sources,
+		       struct sof_sink **sinks, int num_of_sinks)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *source;
+	enum sof_ipc_frame fmt;
+
+	comp_info(mod->dev, "ctc_prepare()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	fmt = audio_stream_get_frm_fmt(&source->stream);
+	if (fmt == SOF_IPC_FRAME_S16_LE)
```

Comment:
use a switch case here, also your catch all should fail in case new formats are added